posts/2023-11-19-piping_data_tables-elio_campitelli/index.qmd

```diff
@@ -0,0 +1,152 @@
+```
+
+Personally, I don't like this syntax very much.
+No matter how you slice it, you always get what feels to me as incomplete lines.
+Also, RStudio doesn't correctly indent the second syntax automatically.
+
+## data.table and magrittr
+
+Until the introduction of the native pipe, I used to write long data.table pipelines using magrittr.
+To do this, I took advantage of the **`.`** placeholder which, within a magrittr pipe, refers to the result of the previous step.
+
+```{r}
+library(magrittr)
+
+penguins[species == "Chinstrap"] %>%
+  .[ , .(mean_flipper_length = mean(flipper_length_mm)), by = .(sex, island)]
+```
+
+I really like this syntax as it's very clean.
+Each line of code is a complete operation without dangling parts and it's easy to comment out single steps.
+
+The only downside is that the dot here has two meanings: as the placeholder for the previous result in `.[`, and as an alias for list in `.(mean_flipper_length = mean(flipper_length_mm))`.
+It's not a huge issue, though, since I tend to read `.[` as a single entity, but it can trip up some people.
+
+## Using the native pipe
+
+The native pipe at first didn't have a placeholder and it didn't chaining to `[`, so this so the above syntax wasn't directly applicable.
+But you could cheat by creating an alias for `[` and use that alias as a regular function.
+So this works:
+
+```{r}
+DT <- `[`
+
+penguins[species == "Chinstrap"] |> 
+  DT( , .(mean_flipper_length = mean(flipper_length_mm)), by = .(sex, island))
+```
+
+This worked so well that data.table officially added the `DT()` function (currently only in the development version), so if you're using the latest development version you don't even need the first line[^2].
+
+[^2]: This function also allows the user to use data.table syntax to data.frames and tibbles retaining the class of the output.
+
+This syntax is fine but I don't like that I need ro write one more character and the closing character being a `)` can get confusing because it adds to the closing `)` that you usually have in the `by` argument.
+
+From R 4.3.0 onwards, the native pipe supports a `_` placeholder to the right-hand side fo the pipe.
+So now the magrittr syntax can be directly translated to
+
+```{r}
+penguins[species == "Chinstrap"] |> 
+  _[ , .(mean_flipper_length = mean(flipper_length_mm)), by = .(sex, island)]
+```
+
+I like this syntax even more than the original magrittr one because it solves the double meaning problem and operations get hugged by a pair of brackets.
+
+## The four pipes of data.table
+
+So, there you are, 4 different ways you can pipe your data.tables.
+
+Use the `][` pipe if you want your code to have minimal dependencies and work in older versions of R.
+Use the `%>%` pipe if you want your code to work in older versions of R and don't mind the extra dependency.
+Use any version of the `|>` pipe if you want minimal dependencies and don't mind depending on R \>= 4.3.0.
+
+```{r}
+penguins[species == "Chinstrap" ][
```

Comment:
This is great thanks for the contribution @eliocamp 
You are right that the automatic code indentation can be tricky.
I usually prefer the version below which seems to indent well at least, but it does not have the nice property about commenting individual operations. That is definitely a great advantage to the pipe, which I had not previously appreciated.
```
penguins[
  species == "Chinstrap"
][ , .(mean_flipper_length = mean(flipper_length_mm)), by = .(sex, island)]
```
Finally, it may be worth mentioning that both piped operations can be done in a single square brackets?
```r
penguins[
  species == "Chinstrap",
  .(mean_flipper_length = mean(flipper_length_mm)), 
  by = .(sex, island)
]

Reply:
I remember there was a thread in the data.table repo about different strategies, but I couldn't find it. I'll add that format too. 

You're right that the example can be done in a single combined operation. Maybe I need to think of one that can't.

Reply:
and actually it can be commented, right? did you mean like this?
```r
DT[
  ## operation 1
][
  operation 2
]

Reply:
Yep, you're right. I updated the post to add that version.